File: Top_100_Liked_Questions/647._Palindromic_Substrings/solution.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        # expand around center method - constant space
        ans = 0
        for center in range(len(s)):
            for i in range(2):
                left = center
                right = center + i
                if right == len(s):
                    break
                while left >= 0 and right < len(s) and s[left] == s[right]:
                    ans += 1
                    left -= 1
                    right += 1
        
        return ans
                    
    
    def countSubstrings2(self, s: str) -> int:
        # dp using memoization
        # a string is palindrome if the first and last characters are same and the remaining substring is a palindrome
        mem = [[False] * len(s) for _ in range(len(s))]
        
        for i in range(len(s)-1):
            mem[i+1][i] = True
        
        ans = 0
        
        for leftIdx in range(len(s)-1,-1,-1):
            for rightIdx in range(leftIdx, len(s)):
                if leftIdx == rightIdx or (mem[leftIdx+1][rightIdx-1] and s[leftIdx] == s[rightIdx]):
                    mem[leftIdx][rightIdx] = True
                    ans += 1
                    
        return ans

File: Top_100_Liked_Questions/647._Palindromic_Substrings/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var countSubstrings = function(s) {
    let ans = 0;
    for (let center=0; center<s.length*2-1; center++) {
        let left = Math.floor(center / 2);
        let right = left + center % 2;
        
        while (left >= 0 && right < s.length && s[left] == s[right]) {
            ans++;
            left--;
            right++;
        }
    }
    
    return ans;
}

var countSubstrings2 = function(s) {
    // dp with memoization
    let mem = [];
    for (let i=0; i<s.length; i++) {
        let row = new Array(s.length).fill(false);
        mem.push(row);
    }

    for (let i=0; i<s.length-1; i++) {
        mem[i+1][i] = true;
    }
    
    let ans = 0;
    for (let left=s.length-1; left>-1; left--) {
        for (let right=left; right<s.length; right++) {
            if (left == right || (mem[left+1][right-1] && s[left] == s[right])) {
                ans++;
                mem[left][right] = true;
            }
        }
    }

    return ans;
};